doc/rfc001-sum-types.md

```diff
@@ -0,0 +1,846 @@
+\Gamma, x_1 \vdash e_1 : (t, c_1) \quad \ldots \quad \Gamma, x_n \vdash e_n : (t, c_n) \quad
+\Gamma, \langle v \rangle \vdash e_{n+1} : (t, c_{n+1}) \quad
+fresh(v)
+}{
+\Gamma \vdash \ match \ e \ \\{ i_1 : x_1 \Rightarrow e_1, \ldots, i_n : x_n \Rightarrow e_n \\} : (t, 
+c \land c_1 \land \ldots \land c_n \land c_{n+1} \land
+s \sim \langle i_1 : t_1, \ldots, i_n : t_n | v \rangle)
+}
+$$
+
+Compared with quint's rule for product construction we see the same
+complementary symmetry. However, we also see a striking difference: there is no
+row variable occurring in the product construction, but the row variable plays
+an essential function in sum type elimination of row-based variants. Row types
+in records are useful for extension operations (i.e., which we don't support in
+quint currently) and for operators that work over some known fields but leave
+the rest of the record contents variable. But the core idea formalized in a
+product type is that the constructor _must_ package all the specified things
+together so that the recipient _can_ chose any thing; thus, when a record is
+constructed we must supply all the things and there is no room for variability
+in the row. For sum types, in contrast the constructor _can_ supply any one thing (of
+a valid alternate type), and requires the recipient _must_ be prepared to
+handle every possible alternative. 
+
+In the presence of row-polymorphis, however, the responsibility of the recipient
+is relaxed: the recipient can just handle a subset of the possible alternatives,
+and if the expression falls under a label they are not prepared to handle, they
+can pass the remaining responsibility on to another handler.
+
+Here is a concrete example using the proposed syntax, note how we narrow the
+type of `T`:
+
+```quint
+type T = A | B;;
+def f(e) = match e {
+  | A => 1
+  | B => 2
+  | _ => 0
+}
+
+// f can be applied to a value of type T
+let a : T = A
+let ex : int = f(A)
+
+// but since it has a fallback for an open row, it can also handle any other variant
+let foo = Foo
+let ex_ : int = f(foo)
+```
+
+Here's the equivalent evaluated in OCaml as proof of concept:
+
+```ocaml
+utop # 
+type t = [`A | `B]
+let f = function
+  | `A -> 1
+  | `B -> 2
+  | _  -> 0
+let ex = f `A, f `Foo
+;;
+type t = [ `A | `B ]
+val f : [> `A | `B ] -> int = <fun>
+val ex : int * int = (1, 0)
+```
+
+## Dynamics
+
+TODO
+
+
+## Concrete Syntax 
+
+Other languages with polymorphic variants:
+
+- ReScript: https://rescript-lang.org/docs/manual/latest/polymorphic-variant
+- OCaml: https://v2.ocaml.org/manual/polyvariant.html
+
+Considerations
+
+- Assuming we support anonymous variant types, we need a way of constructing
+  variants without pre-defined constructors. Potential approaches include:
+  - A special syntax that (ideally) mirrors the syntax of the type 
+  - A special lexical marker on the labels (what ReScrips and OCaml do), e.g., 
+  - Reserve uppercase letters for variant injectors
```

Comment:
Now thinking about this, I would prefer not to have anonymous/open variant types. We have them in Apalache, as we had to introduce ad hoc type annotations. However, the very first thing I always have to do in Apalache is to write a few constructors by hand that eliminate polymorphism. For example, see [erc20_mempool_typedefs.tla](https://github.com/informalsystems/tla-apalache-workshop/blob/f327c96e58858e585a7430e820eead97b668c344/examples/erc20/erc20_mempool_typedefs.tla#L18-L36).

This behavior in Apalache is quite annoying. It causes duplication: (1) Declare the variant type; (2) Introduce constructors by hand. I would prefer to have Rust-like `enum` type declarations that implicitely produce constructors.